[CORL-1434] Optionally show flattened replies

nick-funk opened the pull request:
Superseded by https://github.com/coralproject/talk/pull/3327

---

To test in watch mode, add the following to your `src/core/client/stream/index.html` inside your `head` tag:

```
<script type="application/json" id="config">
      {"staticURI":"/","featureFlags":["FLATTEN_REPLIES"]}
</script>
```

Also, enable `FLATTEN_REPLIES`.

nick-funk committed:
Create initial flattenedReplies resolver and container

- Adds a flattenedReplies resolver implementation
- Creates an empty FlattenedReplyListContainer
- Creates the pagination query for the new container
targeting the new resolver

CORL-1434

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -446,6 +446,22 @@ export const retrieveCommentRepliesConnection = (
+    filter: {
+      ...input.filter,
+      storyID,
+      ancestorIDs: [parentID],
```
```suggestion
      ancestorIDs: parentID,
```

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -3218,6 +3218,12 @@ type Comment @cacheControl(maxAge: 5) {
+    first: Int = 10
+    orderBy: COMMENT_SORT = CREATED_AT_ASC
+    after: Cursor
+  ): CommentsConnection!
```
Suggestion to instead add a `flattened: Boolean = false` on the `replies` edge instead.

nick-funk committed:
Move flattening logic into the replies resolver

CORL-1434

nick-funk committed:
Feed in optional flatten replies variable

Feeds it in as an optional parameter for queries,
fragments, mutations, and subscriptions. We set it
to true for now and will pull from the local store
and the settings object later.

CORL-1434

nick-funk committed:
Create new FLATTEN_REPLIES feature flag

Use it to conditionally load the flattened replies
in subscriptions, mutations, and queries.

CORL-1434

nick-funk committed:
Inject featureFlags into StaticConfig

CORL-1434

nick-funk committed:
Show CommentContainer's for flattened comments

CORL-1434

nick-funk committed:
Make flattened replies collapsible

CORL-1434

nick-funk committed:
Update reply creation to work with flattened replies

CORL-1434

nick-funk committed:
Hook up show all button on flattened replies

CORL-1434

nick-funk committed:
Make 3rd indented replies show in flattened replies

CORL-1434

nick-funk committed:
Hook up view new mutation for flattened replies

CORL-1434

nick-funk committed:
useLocal to check feature flags in commenter stream

CORL-1434

nick-funk committed:
Properly evaluate flatten replies query conditional

CORL-1434

nick-funk committed:
Set id, data-testid on flattened replies container

CORL-1434

nick-funk committed:
Make subscriptions work for flattened replies

CORL-1434

nick-funk committed:
Update all tests except for CommentContainer

CORL-1434

alinhle committed:
fix: restore CommentContainer test

nick-funk committed:
Merge remote-tracking branch 'origin/next' into feature/CORL-1434-flattened-replies

# Conflicts:
#	src/core/client/stream/tabs/Comments/Comment/ReplyCommentForm/CreateCommentReplyMutation.ts

tessalt commented on src/core/client/stream/tabs/Comments/FlattenedReplyList/FlattenedReplyContainer.tsx:
```diff
@@ -0,0 +1,76 @@
+}
+
+const FlattenedReplyContainer: FunctionComponent<Props> = ({ comment }) => {
+  return <div></div>;
```
why the empty component?

alinhle replied:
Looks like you forgot to remove this file

nick-funk replied:
Because this component is never used and should be deleted! I'll do that now, thanks for catching that!

tessalt commented on src/core/client/stream/tabs/Comments/ReplyList/ReplyList.tsx:
```diff
@@ -79,6 +80,11 @@ const ReplyList: FunctionComponent<ReplyListProps> = (props) => {
                       onRemoveAnswered={props.onRemoveAnswered}
                       toggleCollapsed={toggleCollapsed}
+                      ancestorID={
+                        props.flattenLastReply && props.indentLevel === 3
```
the `indentLevel` stuff feels a bit magic-number-y. maybe extract into a well-named constant or add comments to explain

nick-funk replied:
👍

tessalt commented on src/core/client/stream/tabs/Comments/Stream/AllCommentsTab/CommentEnteredSubscription.tsx:
```diff
@@ -99,10 +99,70 @@ function determineDepthTillAncestor(
 }
 
+function getFlattenedReplyAncestorID(comment: RecordProxy, depth: number) {
+  const iterations = depth - 3;
```
yeah once again the 3 here feels arbitrary

nick-funk replied:
👍

tessalt reviewed (changes requested):
couple small issues

alinhle commented:
It crashes when replying to a comment at the level just before the last level.
![image](https://user-images.githubusercontent.com/14221600/100274673-22f7dd00-2f5f-11eb-8d2a-6412aaf92b7d.png)

alinhle commented on src/core/client/stream/tabs/Comments/FlattenedReplyList/FlattenedReplyListContainer.tsx:
```diff
@@ -0,0 +1,241 @@
+          </HorizontalGutter>
+        ))}
+        {relay.hasMore() && (
+          <Indent level={4} noBorder>
```
I think i'd be very useful to extract the const `4` to a variable.

nick-funk replied:
Agreed, did this in previous fixes requested by @tessalt

alinhle commented on src/core/client/stream/tabs/Comments/Stream/AllCommentsTab/CommentEnteredSubscription.tsx:
```diff
@@ -116,6 +176,12 @@ function insertReply(
     return;
   }
 
+  if (depth >= 4 && flattenLastReply) {
```
Agree the `4` should be extracted into a well named constant.

nick-funk replied:
Agreed, did this in previous fixes requested by @tessalt

nick-funk committed:
Remove unused component

CORL-1434

nick-funk committed:
Store flattened reply indent levels in constants

CORL-1434

nick-funk committed:
Check comment.replies is not null or undefined

CORL-1434

nick-funk committed:
Fix up a few more indent level constants

CORL-1434

nick-funk committed:
Revert "Check comment.replies is not null or undefined"

This reverts commit 70a6d65a9533f9b3979a7db1f5831c2876233c9a.

nick-funk committed:
Remove feature flag override in debug stream index.html

CORL-1434
